jax/_src/lax/control_flow/conditionals.py

```diff
@@ -509,15 +583,51 @@ def augment_jaxpr(jaxpr, res_indices):
     jaxpr_aug = core.ClosedJaxpr(jaxpr_aug, jaxpr.consts)
     return jaxpr_aug
 
-  return tuple(_map(augment_jaxpr, jaxprs, res_aval_indices_per_jaxpr))
+  return tuple(map(augment_jaxpr, jaxprs, res_aval_indices_per_jaxpr))
 
 def _ordered_unique(xs):
   d = collections.OrderedDict((x, None) for x in xs)
   return list(d.keys())
 
+def _cond_dce_rule(used_outputs: List[bool], eqn: core.JaxprEqn,
+                   ) -> Tuple[List[bool], core.JaxprEqn]:
+  if not config.after_neurips:
+    return [True] * len(eqn.params['jaxpr'].in_avals), eqn
+  closed_branches = eqn.params['branches']
+  branches = [closed_jaxpr.jaxpr for closed_jaxpr in closed_branches]
+
+  # First, compute which inputs are used in any branch (not including `pred`).
+  used_inputs: List[bool] = [False] * (len(eqn.invars) - 1)  # -1 for pred
+  for jaxpr in branches:
+    _, used_inputs_ = pe.dce_jaxpr(jaxpr, used_outputs, instantiate=False)
+    used_inputs = map(operator.or_, used_inputs, used_inputs_)
+
+  # Next, compute DCEd branches, instantiating according to used_inputs.
+  dce_branches_ = [pe.dce_jaxpr(jaxpr, used_outputs, instantiate=used_inputs)[0]
+                   for jaxpr in branches]
+  dce_branches = [core.ClosedJaxpr(jaxpr, closed_jaxpr.consts)
+                  for closed_jaxpr, jaxpr in zip(closed_branches, dce_branches_)]
+
+  # Finally, update parameters and form the new eqn.
+  dce_linear = [l for l, used in zip(eqn.params['linear'], used_inputs) if used]
+  new_params = dict(eqn.params, branches=tuple(dce_branches),
+                    linear=tuple(dce_linear))
+  # TODO(mattjj,sharadmv): don't assume effects are never DCE'd?
```

Comment:
Can you not do a join across the branch effects here as well?

Reply:
This was stale, not updated after https://github.com/google/jax/pull/11641. Will fix...

Reply:
Ah actually, we can do a join here but `dce_jaxpr` will never eliminate effects. Really this TODO should go elsewhere, and I'll update the code here to be ready for `dce_jaxpr` potentially eliminating effects.